[WIP] Botany Mech

tralezab opened the pull request:
## About The Pull Request

![image](https://user-images.githubusercontent.com/40974010/56786148-bd54d500-67ac-11e9-8cf0-55f9f48db3f1.png)

OK, so lets just all agree: bureaucracy mech was a **bad** idea. And I'm with you on that. But then this robot looking guy gave me a great idea: Botanic Panic.

This mech is mostly made out of plastic, so it is not spaceproof. The shielding in the front covers your face enough to be unidentifiable, though.

WILL HOPEFULLY INCLUDE:
Funny plastic plonking footsteps
Chest water measure and internal storage
A massive hose for watering many crops, and blasting people away from you with jet water.

## Why It's Good For The Game

This feels like the backpack water tank the botanist gets, as in they get to pick how they are going to accomplish their job. Like the miner who wants to gundam lavaland with a reinforced ripley, the botanists clanking around in their plastic mech hosing down plants is rad. Would I remove the eyeballs of anyone who builds a bureaucracy mech roundstart? The answer is unequivocally yes. But with this, I think it'll turn into something the roboticists love to build roundstart to help the station instead.

### I've finished the front facing sprite, i'll start work on this again if it survives the night unlike my last ill-suited (get it) pr. Any ideas for mech tools are appreciated

## Changelog
:cl:
add: Robotics can now complete the botany mech!
/:cl:

tralezab committed:
icon

tralezab committed:
src, and the beginnings of the hose

tralezab commented:
Here's a closeup of the sprite

![botanymech](https://user-images.githubusercontent.com/40974010/56787761-5f2af080-67b2-11e9-8536-d9c8166e926d.png)

optimumtact commented:
mechs are garbage code and so is this code with these spawns

Im gonna give you a 24 hour window to come up with something that looks like someone who gave a shit about code standards wrote it but after that im going to close it.

necromanceranne commented:
Sad this isn't an organically grown mech. I'd laugh so hard if you could make a tree mech powered by potato batteries and glowcaps wires.

tralezab commented:
I would have really enjoyed that but roboticists dont really have enough early game mecha to work on at the moment, and I like the department cooperation

ATH1909 commented:
> Sad this isn't an organically grown mech. I'd laugh so hard if you could make a tree mech powered by potato batteries and glowcaps wires.

What if, instead of the stock parts that are currently used in mech creation, this botany mech required specific kinds of plants, and the stats of those plants affected the stats of the mech (like how the stock parts of other mechs currently affect the stats of the mechs that are made with them).

But yeah, I'm bummed that there won't be a plant that could grow circuitboards for this mech or something.

You'd better add a coconut gun as a mech module (shoots plants from an internal plant bag-like reserve (that you'd have to load with plants yourself).

tralezab commented:
I'll at least make the circuits grown, so botany isn't just always asking for a mech and then doing their normal route. It should at least cost some kind of time for both sides of the deal

tralezab committed:
aquifer water tank icon, jesus the hose code WAS bad

ExcessiveUseOfCobblestone commented:
is this a real pr

tralezab commented:
Yeah. I have a commit with the hose code and its coming along, removing all the garbage copycode and adding the hydroponic watering. I'm also starting on a weeder that removes and converts the weed into an improved fertilizer

ATH1909 commented:
I kind of feel like the parts to this mech should be hand-craftable with planks (or logs?). 25 planks for each limb, with the torso taking a full stack of 50?

ATH1909 commented:
If we do make the mech a sort of giant plant thing made out of wood, I think it'd be rad if, instead of being able to put MMIs/intellicards into it, you could plant a podperson seed (with blood in it) in it and have the person who would have been grown from that seed become (integrated with) the mech.

What if, instead of letting the mech accept normal mech modules, we made the mech gain special botany mech-only modules (up to a cap, of course) based on the traits of plants that you could feed it after its made (and on the traits of the grown mech board)? A plant with bioluminescence would give it a module that'd serve as an always-on light that doesn't use charge (do note that the mech wouldn't normally have a light (even a toggle-able one) at all), a plant with bluespace activity would give it a module that'd let it blink (with a cooldown) like the wizard spell, a plant with fire resistance would give it a fireproofing/lavaproofing module, a plant with perennial growth would give it the ability to use charge to passively repair itself (like the normal self-repair mech module), etc. The effectiveness of these modules (their power cost, the power of their effects, etc.) would depend on the stats of the plants used to add them.

RaveRadbury commented:
Everyone's got big dreams of wooden Rick & Morty mechs.

PKPenguin321 commented:
Absolutely based

WJohn commented:
but why

tralezab commented:
Roboticists don't have a very static early game mech wise, since all they can build is the ripley. If the botanists want to work with robotics early on, they are exploring a different "route" for botany. I put this in the OP, but the gundam miner is a different way to play and in my humble opinion is much more fun than the traditional way.

tralezab committed:
hose fixes + weed converter WIP

tralezab committed:
icon edits, compile

WJohn commented:
but no seriously why do some of the most idle people on the station need a mech that is made obsolete the second you get earthsblood?

tralezab commented:
they move around for glowshrooms, mutagen, planting in public, I could see it. In the final product it will have some unique rewards (it already turns weeds into improved fertilizer), I was thinking of having a seed launcher that would use water in the system to launch a dirt plot with the seed growing in it for mass planting of plots. Finally, I may even consider some benefit from growing in dirt plots to make the mech have to get around more and not get godblooded 

I mean PKP really wanted it he probs has some great ideas

ATH1909 commented:
https://www.youtube.com/watch?v=A7OnzDEQMwQ

habader commented:
make it able to plant kudzu

knacker48 commented:
> make it able to plant kudzu

Make it FROM kudzu

ATH1909 commented:
> > make it able to plant kudzu
> 
> Make it FROM kudzu

Actually, that's not a half-bad idea- kudzu mutations are harder to collect onto a single plant seed than normal plant traits (most of which can just be genemodded onto a single seed using the traits of seeds available from the vending machines). It also gives botany a non-malicious excuse to grow kudzu, which'll be fun.

tralezab commented:
if we REALLY want a giant treant mech then i'll take a look at redoing the flavor but the tools the treant mech uses stays as something robotics has to do.

habader commented:
> if we REALLY want a giant treant mech then i'll take a look at redoing the flavor but the tools the treant mech uses stays as something robotics has to do.

we REALLY want a giant tree mech,
it'd be better than _this_ crappy plastic toy

tralezab commented:
ok, done. i've swapped my fun little plastic mech with a wooden treant. if you don't have enough potency though, you're going to get the really lame treant instead. treat your murmur wood nicely!

tralezab committed:
mostly seeds, but also a lot of mecha work

tralezab commented on code/modules/hydroponics/grown/misc.dm:
```diff
@@ -157,6 +157,26 @@
+	icon_state = "seed-murmuring"
+	species = "murmuring"
+	plantname = "Murmuring Tree"
+	product = /obj/item/reagent_containers/food/snacks/grown/shell/gatfruit
```
i know, i know. this needs to change to the actual mech itself.

FantasticFwoosh replied:
Just a helpful pointer, [the application of .addjuice can help you modify the effects of your plant based on potency to achieve what you spoke about if you wish to use my own past attempt as a example](https://github.com/tgstation/tgstation/pull/26751/files#diff-a54f04b9463de9207e825007965b42ebR100). 

> if you don't have enough potency though, you're going to get the really lame treant instead

tralezab replied:
Hey, great pointer. Right now it drops a big cocoon you smack open for the mech, I can have that be similar to what you linked

FantasticFwoosh commented:
E.N.T Mech (Energetic Non-electric Tree)

tralezab committed:
this pleases the nut

(adds the giant shell from harvesting the murmuring tree. it must be broken open, and it has a mech inside!

tralezab committed:
this compiles the nut

ATH1909 commented on code/game/mecha/mecha_wreckage.dm:
```diff
@@ -197,6 +197,10 @@
+/obj/structure/mecha_wreckage/aquifer
+	name = "\improper Aquifer wreckage"
+	icon_state = "aquifer-broken"
+	desc = "Return to nothing... Just kidding, this shit is plastic."
```
The aquifer is no longer made of plastic...

My immersion is RUINED, smh...

PKPenguin321 replied:
Yknow don't, say swears

tralezab replied:
Right. I'm sorry.

ATH1909 commented on code/game/mecha/working/tree.dm:
```diff
@@ -0,0 +1,37 @@
+	max_integrity = 150
+	lights_power = 0
+	deflect_chance = 0
+	armor = list("melee" = 25, "bullet" = 20, "laser" = 0, "energy" = 0, "bomb" = 40, "bio" = 0, "rad" = 20, "fire" = 100, "acid" = 100)
```
Why is a wooden mech fireproof?

tralezab replied:
left over from plastic mech (which probably should not have been fireproof or acidproof either!)

ATH1909 commented on code/game/mecha/working/tree.dm:
```diff
@@ -0,0 +1,37 @@
+	max_integrity = seed.potency*15
+	obj_integrity = max_integrity
+	if(seed.potency > 5)
+		max_equip += 2
```
Could you make the max_equip bonus be equal to [yield/5] (rounded down)?

tralezab replied:
good idea considering yield is now useless

ATH1909 commented on code/game/mecha/working/tree.dm:
```diff
@@ -0,0 +1,37 @@
+		seed = new /obj/item/seeds/murmuring_wood()
+	seed.forceMove(src)
+	create_reagents(seed.potency*10)
+	max_integrity = seed.potency*15
```
The integrity of the mech should be based on the plant's health stat.

Maybe we could let production speed influence the speed of the mech?

Not sure how we'd make the lifespan stat affect the mech... Chance to resist an EMP? The highest lifespan stat you can get without admin intervention is around 60, IIRC, which sounds like a good maximum chance to resist an EMP blast.

PKPenguin321 replied:
This will allowthem to have way too much endurance because botany

tralezab replied:
You can still change the stats with the gene modder, you just can't add/remove traits.

ATH1909 commented on code/modules/hydroponics/plant_genes.dm:
```diff
@@ -394,6 +394,10 @@
 
+/datum/plant_gene/trait/settled_traits
+	//The traits cannot be changed on this plant. Would make for a pretty funny traitor item, because you could give this but not take it away.
+	name = "Settled Traits"
```
Will this appear randomly in strange seeds?

tralezab replied:
i should fix that

ATH1909 commented on code/modules/hydroponics/grown/misc.dm:
```diff
@@ -157,6 +157,53 @@
+	yield = 1
+	potency = 60
+	growthstages = 3
+	rarity = 60 // Just stick to buying it from cargo, OK?
```
So will this only be acquirable from cargo, or will it be an expensive item in seed vending machines?

tralezab replied:
Probably both, the first one is expensive in the machine and every after is cargo only

ATH1909 commented:
This PR is looking really good so far.

Should the seeds for this mech be added to the pool of seeds that can spawn in the crates in the lifebringer ship? If so, maybe we _should_ add a way to add modules (that are bad/not very good in comparison to mechanical modules, so that requesting parts from robotics is still optimal) to the mech by using plants with certain traits on it?

optimumtact commented:
If you are not a maintainer and not commenting on specific code issues, and instead are suggesting design changes or features, can you please refrain from using the code review comments system

ATH1909 commented:
Okay, sorry.

So you can still edit the stats of these seeds? Then we definitely should make the mech's health be based on endurance, as that's a stat that's much harder to boost up to 100 than potency.

So the benefits from each stat should ideally look like this?
potency: increases maximum reagent capacity of the mech and maybe either the effectiveness of the plant mech-unique modules or the power efficiency of the mech? I dunno
lifespan: % chance to negate (or at least reduce) the effects of an EMP pulse on the mech
endurance: increases maximum integrity of the mech
production speed: gives the mech extra speed (should probably make the base speed of the mech slower to accommodate for this)
yield: gives a number of extra module slots to the mech
the other stats: no one cares about them (except for grass's slightly faster than normal maturation rate, ambrosia gaia's much worse weed vulnerability stat, etc., but they're not really very relevant to the plant mech)

tralezab commented:
well, I wanted the mech to be accessable early, but if you waited late game it would be a lot better. I may make the other stats give bonuses if they are high, but in the end potency will buff all other stats and matter the most.

habader commented:
Tbh I'd just skip out all stats other than potency

tralezab commented:
This is not left in the dust I have been very busy and the implications of a non mechanical mech are aplenty so work has been slow

tralezab committed:
compiles + removing lights action

tralezab commented:
needs a lot of work and fluff to feel right and a ton of sounds, and the reception of a botany mech will probably not be pretty. i'll make something a little more sane that doesn't have to do with a mech

ATH1909 commented:
F
